src/main/java/io/fabric8/jenkins/openshiftsync/BaseWatcher.java

```diff
@@ -0,0 +1,118 @@
+import static java.util.logging.Level.SEVERE;
+import hudson.triggers.SafeTimerTask;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.openshift.api.model.BuildConfigList;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.csanchez.jenkins.plugins.kubernetes.PodTemplate;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import jenkins.util.Timer;
+
+public abstract class BaseWatcher {
+    private final Logger LOGGER = Logger.getLogger(BaseWatcher.class.getName());
+
+    protected ScheduledFuture relister;
+    protected final String[] namespaces;
+    protected Map<String,Watch> watches;
+
+    @SuppressFBWarnings("EI_EXPOSE_REP2")
+    public BaseWatcher(String[] namespaces) {
+        this.namespaces = namespaces;
+        watches = new HashMap<String,Watch>();
+    }
+    
+    public abstract Runnable getStartTimerTask();
+
+    public synchronized void start() {
+        // lets do this in a background thread to avoid errors like:
+        // Tried proxying
+        // io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration to support
+        // a circular dependency, but it is not an interface.
+        Runnable task = getStartTimerTask();
+        relister = Timer.get().scheduleAtFixedRate(task, 100, 10 * 1000,
+                TimeUnit.MILLISECONDS);
+    }
+    
+    public synchronized void stop() {
+        if (relister != null && !relister.isDone()) {
+          relister.cancel(true);
+          relister = null;
+        }
+
+        for(Map.Entry<String,Watch> entry:watches.entrySet()) {
+          entry.getValue().close();
+          watches.remove(entry.getKey());
+        }
+      }
+
+    public synchronized void onClose(KubernetesClientException e) {
+        if (e != null) {
+            LOGGER.warning(e.toString());
+
+            if (e.getStatus() != null && e.getStatus().getCode() == HTTP_GONE) {
+                stop();
+                start();
+            }
+        }
+    }
+    
+    protected PodTemplate podTemplateInit(PodTemplate podTemplate) {
+        //podTemplate.setInstanceCap(Integer.MAX_VALUE);
+        podTemplate.setCommand("");
+        podTemplate.setArgs("${computer.jnlpmac} ${computer.name}");
+        podTemplate.setRemoteFs("/tmp");
+        File f = new File("/run/secrets/kubernetes.io/serviceaccount/name");
+        if (f.exists()) {
+            ObjectInputStream ois = null;
+            try {
+                FileInputStream fis = new FileInputStream(f);
+                ois = new ObjectInputStream(fis);
+                String saName = ois.readUTF();
+                podTemplate.setServiceAccount(saName);
```

Comment:
I'd really like to avoid 1) and 2)

I'll look into 3) but I remembered last night that the login plugin is able to get the service account name using the service account token and the `/oapi/v1/users/~` endpoint.  I'm going to see if I can leverage that using the sync plugin's fabric8 client.